job-manager: support job annotations by scheduler

chu11 opened the pull request:
This is just a starting point for annotation support in the job-manager.  No support in `job-info` yet, as that is the far harder part of the work.

I would consider this PR merge-able as a stepping point, but we don't have to.  It can wait till bigger work later on comes.

- alloc "note" was previously dropped, now we store it
- add some simple testing support by allowing it to be listed via the `job-manager` list function
- minor cleanup, add some enums in place of some "magic numbers"

one piece of testing coverage missing is for multiple notes to be stored for multiple pending jobs, but AFAICT both `sched-simple` and `sched-dummy` stop attempting to allocate after the next job in the queue can't be scheduled.    Tweaking `sched-dummy` to cover this seems like the best path on this, but pondering it a bit.

garlick commented on src/common/libschedutil/alloc.h:
```diff
@@ -16,6 +16,13 @@
+    FLUX_SCHED_ALLOC_DENIED  = 2,
+    FLUX_SCHED_ALLOC_CANCEL  = 3,
+};
+
```
Nice cleanup, thanks!

chu11 commented:
> Looks good. My only suggestion would be that it seems just as easy to accept a json object here as a string. May as well do that and then after this PR, fluxion could go ahead and provide it?

You mean just check that the string is a json object?  That seems like a good idea.  The idea was to figure out a json format later on, but just validating it as a json object would be good to get some structure in now.

dongahn commented:
> You mean just check that the string is a json object? That seems like a good idea. The idea was to figure out a json format later on, but just validating it as a json object would be good to get some structure in now.

If that's added, it will take me only a few hours for fluxion to produce JSON for testing.

garlick commented:
I mean instead of note being a `s?:s`, make it a `s?:o`.   The note should be a JSON dictionary, and then instead of replacing the previous note (if any), merge with e.g. `json_object_update` so sched can choose to update any of the keys it wants and leave others alone.

At least that was my understanding of the agreed-upon direction.

grondo commented:
If you allow the annotation to be a JSON object, and store it for retrieval by `flux-jobs` it seems like we could trivially support printing members of the object with dotted notation,e .g. `flux jobs -o '{scheduler.estimate}: {scheduler.reason}'`

chu11 commented:
> If you allow the annotation to be a JSON object, and store it for retrieval by flux-jobs it seems like we could trivially support printing members of the object with dotted notation,e .g. flux jobs -o '{scheduler.estimate}: {scheduler.reason}'

yeah, this is what I was thinking ;-)

chu11 commented:
> I mean instead of note being a s?:s, make it a s?:o. 

Right now the other alloc responses (i.e. `R`, `denied`) pass notes as strings.  Are you saying you want to make the "note" a json object universally?

If yes, perhaps we should punt that to a "part 2" PR which can be my follow up to this one.

chu11 commented:
setting WIP so i can debug travis racyness - I'm assuming the annotations haven't propagated to the job-manager in a timely manner.

garlick commented:
> Right now the other alloc responses (i.e. R, denied) pass notes as strings. Are you saying you want to make the "note" a json object universally?

No, I was only suggesting we change the one response type.  Maybe it should have a new name to avoid confusion though.

chu11 commented:
heh, it ends up that `cut -f 6-` isn't working in travis.  So in RHEL7 that gets me the "no cores available" string but in travis it only grabs "no".  Dunno about nuances of different versions. Oh well, will figure out a new way once I update the note to be a json object.

chu11 commented:
to my dismay, there was a portability issue between bash and dash regarding the use of local variables in a function.  I don't fully understand it, but figured out that

```
func() {
    local tmp=$(echo 1 2 3 4 5)
    echo $tmp
}
foo=`func`
echo $foo
```

behaved differently on my local ubuntu box and rhel7 bash.  The fallout behavior in my tests was different (tmp would return a different string) ... but ugh ...

grondo commented:
Hahah first accepted answer in [this related stackoverflow post](https://stackoverflow.com/questions/18597697/posix-compliant-way-to-scope-variables-to-a-function-in-a-shell-script) starts with:

> If you'd like to journey down to Hell with me

I think that about sums it up!

chu11 commented:
I've removed the WIP, I'll have a follow up PR about converting the "annotation" from a string to an object instead of here.  So this PR can be considered independently of the other PR.

grondo commented:
It feels like the two PRs should just be combined, so that a single, whole feature is added in a single PR.

chu11 commented:
> It feels like the two PRs should just be combined, so that a single, whole feature is added in a single PR.

Heh, ok.  If that's the group preference, I'll re-merge back into this PR.

grondo commented:
> Heh, ok. If that's the group preference, I'll re-merge back into this PR.

Well, it was only my personal opinion. However, if you think of a single PR as release note entry, it makes sense that PRs follow rules like

 * Adds a feature
 * Fixes one bug (or more than one, but related bugs)
 * Is cleanup or other refactoring that doesn't change any functionality

I don't feel really strongly about it in this case though.

chu11 commented:
just following up conversations from the con-call, decision was to go with json objects for all notes.  Obviously, only the "annotation" will allow json objects to be updated.

garlick commented:
Did you have a combined PR ready to go yet @chu11?  I have a window available today to do some review.  A couple of thoughts, without having seen the JSON version of this:
* We should probably consider these annotations to be "best effort" in terms of error handling.  Log any concerning errors at LOG_ERR level, but terminating the scheduler interface may be excessive force.
* A new `json_t *alloc_note` dictionary in the internal `struct job` should probably allow for each alloc response to update, not replace, the object.  In other words, use `json_object_update()`
* May as well have all the responses provide the optional note as a dict and update with same semantics.
* When an event is posted in response to an alloc response (all but the NOTE type response), maybe the whole `job->alloc_note` dictionary should be posted in the event context?  That way the final state of the dictionary is persistently logged.

garlick commented:
> Obviously, only the "annotation" will allow json objects to be updated.

Oh I missed this comment.

Did you mean response types other than NOTE would replace not update the dict?  Is it obvious?  I was thinking they all should update just to avoid special cases but maybe there is a use case you had in mind?  (I admit I'm just coming up to speed here and may be missing something obvious!)

chu11 commented:
@garlick nah, first attempt with #2990 we decided to go in another direction, so working on new direction.

Generally agree with all your points.

>   A new json_t *alloc_note dictionary in the internal struct job should probably allow for each alloc response to update, not replace, the object. In other words, use json_object_update()

To clarify, there should be a single note object, and all responses can update it?

> Did you mean response types other than NOTE would replace not update the dict? Is it obvious? I was thinking they all should update just to avoid special cases but maybe there is a use case you had in mind? (I admit I'm just coming up to speed here and may be missing something obvious!)

The reason it was "obvious" to me is b/c only a single `denied` or `R` response can happen.  So there is no "update" there logically.  The "update" only makes sense, per your comment above, that we have a single note object that all responses update.

garlick commented:
> To clarify, there should be a single note object, and all responses can update it?

That's what I was thinking.  Scheduler would effectively own the dict and could put whatever it wants in there, from any of the responses.

We might define a couple of well known keys, e.g. an estimated start time, and a human readable error message (the latter to go in DENIED exception context).

Idea: include the dict in the `alloc` event context so it can be restored across a restart, and pass it back to the scheduler in the `free` request?  (In case scheduler wants to bookmark the resources allocated to the job or something like that?)

Anyway I hope I'm not too far off base here.  LMK if so.

chu11 commented:
> We might define a couple of well known keys, e.g. an estimated start time, and a human readable error message (the latter to go in DENIED exception context).

We will definitely want to do this, perhaps via RFC in a later PR though.

> Anyway I hope I'm not too far off base here. LMK if so.

Sounds good, although I may want to limit how much goes into this PR before it begins to explode.  We'll see as I keep on working on it.

chu11 commented:
so just did a big push:

- most of the original patches are still there, initially support simple string based annotations
- update all notes to be json object across all alloc responses
- update all notes across all alloc responses to be a shared internal data structure
- alloc notes now update the data structure instead of over writing it
- added a new `sched-annotate` scheduler to test paths and corner cases.  the scheduler is a little hacky, but it's sort of that way to intentionally test weird paths
  - there is a lot of cut & paste with `sched-dummy`, i contemplated putting all of it into `sched-dummy` but it was just different / weird enough I split it out.  I could make a common-lib out of common functions.  That's something to consider (it's some, but not a lot of code).

chu11 commented:
yarg ... the curse of `local` on bash vs dash hits me again.   I'll fixup.

```
t2203-job-manager-dummysched.t:  PASS: N=7   PASS=7   FAIL=0 SKIP=0 XPASS=0 XFAIL=0

ERROR: t2203-job-manager-dummysched.t - missing test plan

ERROR: t2203-job-manager-dummysched.t - exited with status 1

t2204-job-manager-annotatesched.t:  PASS: N=7   PASS=7   FAIL=0 SKIP=0 XPASS=0 XFAIL=0

ERROR: t2204-job-manager-annotatesched.t - missing test plan

ERROR: t2204-job-manager-annotatesched.t - exited with status 1
```

chu11 commented:
hmmm asan failed:

```
 ERROR: t2206-job-manager-bulk-state
===================================

flux-broker: join: Connection timed out: Connection timed out
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'PAIR' socket created at src/zsys.c:471
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'PAIR' socket created at src/zsys.c:472
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'REP' socket created at src/zauth.c:80
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'ROUTER' socket created at overlay.c:385
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'PAIR' socket created at overlay.c:356
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'DEALER' socket created at overlay.c:432
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'PAIR' socket created at module.c:668
E: (flux-broker) 20-06-12 23:01:29 [125273]dangling 'PAIR' socket created at shmem.c:187
flux-broker: src/zsock_option.inc:2906: zsock_set_sndtimeo: Assertion `rc == 0 || zmq_errno () == ETERM' failed.
flux-start: 1 (pid 125273) Aborted
flux-start: 3 (pid 125278) Killed
flux-start: 0 (pid 125272) Killed
flux-start: 2 (pid 125275) Killed
ERROR: t2206-job-manager-bulk-state.t - missing test plan
ERROR: t2206-job-manager-bulk-state.t - exited with status 137 (terminated by signal 9?)
```

Possible I mssed up something in the job-manager, but not much to go on.

grondo commented:
> hmmm asan failed:

I had to fix the after-failure scriptlet for the asan builder in my PR because it wasn't correctly finding all the asan output files. If you pull in that fix from my branch, perhaps you'll get the missing info (though I'm not 100% sure I fixed it)

chu11 commented:
> I had to fix the after-failure scriptlet for the asan builder in my PR because it wasn't correctly finding all the asan output files. If you pull in that fix from my branch, perhaps you'll get the missing info (though I'm not 100% sure I fixed it)

Thanks, pulled it in.  I'm not having luck reproducing on TOSS and some re-runs of the asan builder passed.  Dunno if its racy or if it's just slow and by chance it hit some job-manager tests (thus making me nervous).

grondo commented:
I had to adapt timeouts under ASan, which it seems like I never had to do before.
e.g. something like:
```
test_have_prereq ASAN && TIMEOUT=${bigger_timeout}
```

chu11 commented:
well i've re-run the asan test 7-8 times more and no failures.  unable to reproduce on TOSS systems even w/ `taskset 0x01`.  So unsure if we hit a super low odds race or if travis was just slow and hit something weird.

chu11 commented:
rebased & re-pushed, fixing some conflicts from @grondo's recent PRs

chu11 commented on src/modules/job-manager/alloc.c:
```diff
@@ -232,6 +234,32 @@ int cancel_request (struct alloc *alloc, struct job *job)
+                          (uintmax_t)id);
+        }
+        else {
+            if (!(job->note = json_incref (note)))
```
was just thinking about this.  Is this safe or should this be a `json_copy()`?  B/c we might be updating this, we may want our own copy.  But the message's copy should disappear shortly.

garlick replied:
If this is coming from the message payload, IMHO it should be safe to just take a reference (as `json_object_update()` is doing) because the message is destroyed as soon as the handler returns, in the same thread.

chu11 commented:
rebased and re-pushed with two extra commits.  These commits add a simple streaming RPC service to the job-manager, allowing a client to get a stream of all the notes coming from the job-manager.

The service is simple for the time being, just streaming every note change that occurs for everyjob, 1 id + note per RPC message.  In time this could be updated to be a bulk service, etc.  But I think this is a good small step in this PR to eventually get support into the job-info module.

I think this is a good stopping point for this PR, as the next follow up PR will add this support into the job-info module, then eventually into `flux-jobs`.

chu11 commented:
I've hit a ton of failures in travis with

```
FAIL: t2207-job-manager-wait.t 13 - wait works when job terminated by exception
FAIL: t2207-job-manager-wait.t 17 - wait --all works with no waitable jobs
FAIL: t2207-job-manager-wait.t 18 - wait --all works with one job
FAIL: t2207-job-manager-wait.t 19 - wait --all works with two jobs
FAIL: t2207-job-manager-wait.t 20 - wait --all fails when first job fails
FAIL: t2207-job-manager-wait.t 21 - wait --all fails when second job fails
FAIL: t2207-job-manager-wait.t 22 - wait --all --verbose emits one line per succesful job
  t2207-job-manager-wait.t:  FAIL: N=29  PASS=22  FAIL=7 SKIP=0 XPASS=0 XFAIL=0
ERROR: t2207-job-manager-wait.t - exited with status 1
```

i don't touch this part of the code in my PR, so I'm a bit reluctant to think its something in this PR.  But it's most of the builders.  Perhaps something racy that my changes have triggered?  I was able to reproduce once on my ubuntu box at home.

chu11 commented:
Appears my fixes fixed the travis failures.  Fixed a few other misc things and added a little bit more coverage.  There's a few extra coverages I could add, but it seems like a lot for just tiny coverage (i.e. annotation with a non-json object leads to a non-updated note).

Think this is ready for a review.  Removing WIP.

garlick commented on src/modules/job-manager/job-manager.c:
```diff
@@ -107,6 +145,10 @@ int mod_main (flux_t *h, int argc, char **argv)
     zhashx_set_destructor (ctx.active_jobs, job_destructor);
     zhashx_set_duplicator (ctx.active_jobs, job_duplicator);
+    if (!(ctx.note_requests = zlist_new ())) {
+        flux_log_error (h, "error creating note_requests list");
```
You may need to manually set `errno = ENOMEM` here since `zlist_t` makes no guarantees about errno.

garlick commented on src/modules/job-manager/note.c:
```diff
@@ -0,0 +1,150 @@
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* note - handle not requests & send out updates to new notes
```
typo - "not"

garlick commented on src/modules/job-manager/note.c:
```diff
@@ -0,0 +1,150 @@
+    }
+    if (!(msg_ref = flux_msg_incref (msg)))
+        goto error;
+    if (zlist_append (ctx->note_requests, (void *)msg_ref) < 0)
```
set errno = ENOMEM

garlick commented on src/modules/job-manager/note.c:
```diff
@@ -0,0 +1,150 @@
+        errno = EPROTO;
+        goto error;
+    }
+    if (!(msg_ref = flux_msg_incref (msg)))
```
incref cannot fail

garlick commented on src/modules/job-manager/note.c:
```diff
@@ -0,0 +1,150 @@
+        zlist_destroy (&l);
+    }
+    else
+        flux_log_error (ctx->h, "%s: zlist_dup", __FUNCTION__);
```
set errno = ENOMEM when zlist_dup fails

chu11 commented:
@garlick Thanks for the quick skim.  Rebased and re-pushed, squashing the minor nits above into the patch series.

chu11 commented:
So I began working on the `job-info` module support today and I realized a big problem with the "notes" RPC streaming service.

- we had debated if the notes would be streamed via an RPC OR published via an event OR published within the current state transitions event.

- the primary reason for going with a streaming RPC is b/c notes are not stored anywhere (e.g. eventlog) so there would be no way to "recover" the current set of notes from the job-manager.  So the idea was that an RPC could respond with the "current notes" of active jobs in the job-manager, leading to seemless recovery.

  - secondarily, we didn't want to pile a whole bunch of "notes" into the job state transitions, since that data could get very large very quickly.

- however, the RPC streaming service can race with the job-state transition events.  So a note can arrive before the job-info module is aware of a job's existence.  There's ways to deal with this, but it does lead to complexity that we would prefer not to have (e.g. store notes somewhere on a "wait for job id to exist" list).

Thinking about this.

- should it be a note streaming service per job?  Not for all jobs?
- maybe pass along notes in job state transition, store note changes in eventlog for recovery?

For the purposes of this PR, perhaps adding the streaming service was pre-mature.  So I'm thinking I should nix the last 3 commits in this PR.  Only the prior commits should be considered for review.

garlick commented:
Hmm, two thoughts:
1) Should we just have one streaming service that job-info uses to sync up with job-manager, that includes state transitions?
2) Or would it make sense to just have clients of job info (e.g. `flux jobs`) query the current note directly from the job manager, per job?  (not streaming)  Potentially we could do it in one RPC (send list of jobids, return dict of jobid => note)?

chu11 commented:
rebased & re-pushed, removing the top 4 commits from this PR, all related to the notes streaming RPC service.  I think that + initial job-info support will go into a follow up PR.  We'll let the refactor of annotations into json objects and their updateability be the extent of this part 1 PR.

chu11 commented:
> Should we just have one streaming service that job-info uses to sync up with job-manager, that includes state transitions?

Haven't thought about that, but it seems like a decent idea.  Although it would be a lot of work to refactor in the short term.

Right now, I think `job-info` is the only consumer of the `job-state` events.  Did you see other potential consumers of it?  I had assumed so, so didn't think ditching the `job-states` events was an option.

Other potential issue, can the job-manager be reloaded while the job-info module isn't?  Should any RPC streaming service regularly be "re-connected" by the job-info module if it loses its stream?  Perhaps regularly trying via a timer?  Unsure if this could lead to potential job-info / annotations being lost.  With published events, I don't think this can't happen.

> Or would it make sense to just have clients of job info (e.g. flux jobs) query the current note directly from the job manager, per job? (not streaming) Potentially we could do it in one RPC (send list of jobids, return dict of jobid => note)?

It would perhaps allow us to implement this faster, but would be sort of hackey and not a good long term solution.

As I was thinking about it, one solution I presented above:

> should it be a note streaming service per job? Not for all jobs?

I like this solution, with the exception that it could be a lot of simultaneous streams, which may be not-good.

- it would eliminate races, job-info can request streaming notes for a jobid once it knows of the job-id.  It would always get the most current note that job-manager is aware of.
- the stream can be ended after a job is allocated, since you know there can't be any more notes

grondo commented:
> > Or would it make sense to just have clients of job info (e.g. flux jobs) query the current note directly from the job manager, per job? (not streaming) Potentially we could do it in one RPC (send list of jobids, return dict of jobid => note)?

> It would perhaps allow us to implement this faster, but would be sort of hackey and not a good long term solution.

Just my $0.02, but I don't find this solution hacky. We already allow listing of active jobs directly from the job-manager, and have never stipulated that all job data must go through job-info. (For example,  to query more fine-grained job information flux-jobs will have to go to eventlogs, guest kvs data, etc anyway). However, this idea does raise a couple other questions:

 * why not allow flux-jobs or other frontend utilities to query from a service directly exposed by the scheduler and thus eliminate an unnecessary responses to job-manager and extra state this module needs to keep?
 * is there any use case to display the final state of the scheduler annotations? How would that be accomplished.

> >  should it be a note streaming service per job? Not for all jobs?
> I like this solution, with the exception that it could be a lot of simultaneous streams, which may be not-good.

Yeah, there is already a streaming request open to the scheduler from the job-manager for all outstanding alloc requests. This would be doubling that number of requests. Maybe that would be solved by finally implementing an alloc mode between `single` and `unlimited`?

In the case of multiple job-info modules loaded in an instance, there will be even more outstanding RPCs...

garlick commented:
> why not allow flux-jobs or other frontend utilities to query from a service directly exposed by the scheduler and thus eliminate an unnecessary responses to job-manager and extra state this module needs to keep?

Good point, if the scheduler can be relied upon to be responsive.  I was assuming it might periodically do computationally intensive planning, keeping its reactor thread busy, but fluxion is shaping up otherwise.

That seems like a pretty nice way to go.  What thinks @chu11?

>> should it be a note streaming service per job? Not for all jobs?

> I like this solution, with the exception that it could be a lot of simultaneous streams, which may be not-good.

IMHO a show stopper.

chu11 commented:
> Just my $0.02, but I don't find this solution hacky. We already allow listing of active jobs directly from the job-manager, and have never stipulated that all job data must go through job-info. (For example, to query more fine-grained job information flux-jobs will have to go to eventlogs, guest kvs data, etc anyway).

Now that I've thought about it a bit more, I suppose it's not as bad as I originally thought.

> is there any use case to display the final state of the scheduler annotations? How would that be accomplished?

Good question.  If the data were "cached" in the job-info, it would still be available after the job is done.  
But we don't have a way to recover if the job-info module is restarted.

Another thought.  Are scheduler annotations necessary after a job starts running?  Or only while the job is the SCHED state?  Answer to this question could lead to implementation decision.  i.e. "grab all annotations in the job-manager" vs. "only get annotations for specific job IDs from job-manager".

chu11 commented:
> That seems like a pretty nice way to go. What thinks @chu11?

I think it's a good idea to try.  It'll be for round 2 of this PR series.  I will leave this PR in its current state for review.  If we think the basics of it are solid, we'll then need to ask the sched team to update sched.

grondo commented:
> I was assuming it might periodically do computationally intensive planning, keeping its reactor thread busy, but fluxion is shaping up otherwise.

Here's a counterpoint to that point. If indeed the scheduler annotations are rapidly changing while jobs have outstanding alloc requests, then responding with a new note for each change might actually mean the scheduler has to send more RPCs in some situations. Hard to say if that would be true for a system instance, but for a high throughput workload it might make a difference. (eh, maybe not either way)

chu11 commented:
> We already allow listing of active jobs directly from the job-manager, and have never stipulated that all job data must go through job-info.

Pondered this tonight. another option might be for the `job-info` module to retrieve the data from the `job-manager` on demand.  i.e. if the request from the client asks for annotations for jobs, `job-info` will get the information from the `job-manager` before returning to the user.  Pro: all hidden behind same `job-info` interface.  Con: calling `job-info`, then `job-manager` is "serial" and slower than if it was done in "parallel" via `flux-jobs`.

garlick commented:
In the interest of continuing to move forward, don't we go ahead and get this PR landed with the "note" object stored in the job manager as originally proposed, and let fluxion implement the alloc response payload as originally planned.  Then we can separately discuss the best way to get the notes dict to the `job-info` command, and leave the door open to a more extensive redesign later on that could push it back to a service method in the scheduler.

I'll start a more detailed review of this PR unless there are objections to this plan.

chu11 commented:
> In the interest of continuing to move forward, why don't we go ahead and get this PR landed with the "note" object stored in the job manager as originally proposed, and let fluxion implement the alloc response payload as originally planned. 

Agreed.  How to get the info to `job-info` / `flux-jobs` is for later.

@dongahn do you want to start looking at how to update sched?  Also, we may want to start an RFC for conventions on keys within the annotations.

garlick commented:
Started to review but the first half is just "squash incremental development", since this PR evolved from note as a C string to note as a JSON string to note as a dict.  Could you squash that down so that evolution is no longer in the PR and then I'll start over with a review?  (Meanwhile I'll generally look at the finished product).

garlick commented:
Since we have `note` defined in multiple places as a string (exception context, alloc event context, and alloc SUCCESS and CANCEL responses), it feels to me like we ought to at least introduce a new key name in the alloc response for the one case where it's a dict, and where we haven't used it yet in the schedulers.

Could we leave `note` as a `{s?s}` in the alloc response, and add `annotations` as a `{s?O}`?

Then only call the code in your `update_note()` (minus the string decode) if `annotations` appears in the response?  And only set in `schedutil_alloc_resopond_annotate()`?  (leave respond_R and respond_denied with a C string arg called `note`)?

Earlier I suggested that we treat `note` as a dict in all the responses and update job->note in all the responses, but it looks like we ended up with a compromise to avoid breaking the protocol with sched, where `note` is still expected to be a string in SUCCESS and DENIED responses.  It's reasonable to avoid breaking the protocol, but if we're going to go that way, I think we should use a different key name for the dict.

Does that make sense?

chu11 commented:
> Earlier I suggested that we treat note as a dict in all the responses and update job->note in all the responses, but it looks like we ended up with a compromise to avoid breaking the protocol with sched, where note is still expected to be a string in SUCCESS and DENIED responses.

I'm not sure I follow.  the note responses to success and denied should also be json dicts.  This will break protocol and `sched` will have to be updated as a result.

dongahn commented:
> @dongahn do you want to start looking at how to update sched? Also, we may want to start an RFC for conventions on keys within the annotations.

Sounds good to me. Will get to this soonish.

garlick commented:
> I'm not sure I follow. the note responses to success and denied should also be json dicts. This will break protocol and sched will have to be updated as a result. Is some part of the code suggesting otherwise?

Oh - I think the way the code is now, a string `note` would pass the unpack (sender sends `s:s`, receiver parses with `s?o` - which would return a json_t for which `json_is_string()` would be true), but I guess it would then fail when passed  as the second arg to `json_object_update()`?   I missed that.

Also, the job->note dict is encoded to a string and passed as the `note` value to `event_job_post_pack()` in the DENIED and SUCCESS cases.  Isn't that going to wind up as json within a json string?  Was that intended or am I misreading again?

chu11 commented:
I think I may see the confusion.  RPC wise, everything should be passed/decoded as `s:o` (or `s:O`).

However, the schedutil alloc functions take the note as a string, b/c we don't want those functions to take a `json_t`.  The encoded string is verified to be a valid json object, then sent as `s:o`.

So is the confusion that you're seeing things like this?

```
        const char *debug_note="{\"msg\":\"DEBUG_FAIL_ALLOC\"}";
        if (schedutil_alloc_respond_denied (sc->schedutil_ctx,
                                             sc->job->msg,
                                             debug_note) < 0) {
             flux_log_error (sc->h, "schedutil_alloc_respond_denied");
            goto done;
        }
```

Perhaps there should be `vpack()` equivalent to the scheduilt alloc functions to make it clearer what's going on?

Also, I now realize code like this:

```
        if (event_job_post_pack (ctx->event, job, "alloc",
                                 "{ s:s }",
                                 "note", note_str ? note_str : "") < 0)
            goto teardown;
```

Isn't necessary.  I could do `s:o` to `event_job_post_pack()`.  So perhaps that was part of the confusion.

garlick commented:
Or this
```json
{"timestamp":1592940715.1500163,"name":"exception","context":{"type":"alloc","severity":0,"userid":-1,"note":"{\"msg\":\"unsatisfiable request\"}"}}
```

I would think in those cases a note with a msg inside is not as clear as a note that is the message.

chu11 commented:
Just to double check, you're thinking:

> {"timestamp":1592940067.4539638,"name":"alloc","context":{"note":"{\"msg\":\"rank0/core0\"}"}}

into

`{"timestamp":1592940067.4539638,"name":"alloc","context":"{\"msg\":\"rank0/core0\"}"}`

Basically put contents of the note into the context?

I think this would be ok, but one of the reasons to store the entire note as is is so that it could be re-loaded if necessary.

> But backing up, RFC 21 requires note to be a human readable string in the alloc and exception event contexts. Are we proposing to change that?

This is something I missed and wasn't aware of.  I think we have to change it if the annotation is a json-object and we are updating it with subsequent calls?

garlick commented:
Topic for coffee meeting?

I've lost track - did we have a use case for save/restore of the annotations, because we're not posting events when they are updated during PENDING any of those would be lost.  For the "start time projection" use case, it doesn't seem useful to me to save, but now that we have a generalized dict perhaps we are thinking of other use cases where it is useful?  (I may have even made rumblings in that direction...sorry if I'm going the other way now).

In general my feeling at this point is we should only keep the dict in memory and only update it during PENDING with alloc ANNOTATE responses.  For the other alloc response types, I would vote we stick with the single human readable string that we post to the events, thus making it so we don't need to change RFC 21 or the other code that reads those events.  That is why I was suggesting the  "annotations" (dict) vs "note" (string) name change above.

chu11 commented:
>  For the other alloc response types, I would vote we stick with the single human readable string that we post to the events, thus making it so we don't need to change RFC 21 or the other code that reads those events. That is why I was suggesting the "annotations" (dict) vs "note" (string) name change above.

Ok.  I see what you're propsing.  Lets talk during coffee time.    Hopefully @dongahn can make it.

garlick commented:
See also #2608.  That issue mentions user/job annotations which might require persistence.  OTOH, we could make sure that each of those generates an event, whereas the projected start time updates probably should not.

chu11 commented:
FWIW, re-pushed with a bunch of commits squashed, so it's fewer commits overall.

chu11 commented:
During coffee time @garlick and I sort of came to the conclusion that there is a central assumption that really guides the design/implementation of the annotations.  And we need to decide on this point to decent how this PR should be implemented.

The key design point is whether annotations should be readable/queryable when a job is pending OR if it should be readable/queryable after the job is allocated/denied.

If annotations should only be readable/available when a job is pending,  i.e. once a job starts running / is denied, annotations are no longer available to be read, this implies that annotations should only be json objects when the "annotation" response is sent.  The final "note" for a allocation/denied response does not need to be preserved, so it need not be a json object.  The final "note" can simply be a string that gets written to the eventlog, and after that point, no preservation of prior annotations is necessary.

If annotations should be queryable after a job has been allocated resources/denied, it implies "notes" always being json objects, including the notes sent during an allocated/denied response.  Because we want to preserve these notes, the final "note" is a json object that only updates the stored note.  Note that how to preserve across a restart is still TBD and how to `job-info` could cache/load this annotation is TBD.

Hopefully what I'm saying makes sense.

Thoughts on these two approaches:

- My initial goal in this PR was based on the latter assumption.  However, using the first assumption, where we don't care about the annotations after the job is allocated/denied is easier and might be a good "part 1" implementation regardless of what we decide longer out.  The preservation of the annotations after a job is running/denied has several TBDs that we just don't know the answer to yet.  It might be easier/better to solve the easier case first.

dongahn commented:
>> @dongahn do you want to start looking at how to update sched? Also, we may want to start an RFC for conventions on keys within the annotations.
> Sounds good to me. Will get to this soonish.

Looks like there are still churns including RFC updates. I think I should wait a bit before upping fluxion for this. Is this correct @chu11?

chu11 commented:
> Looks like there are still churns including RFC updates. I think I should wait a bit before upping fluxion for this. Is this correct @chu11?

Yup, lets wait until that RFC is done and I push an update given the RFC (working on it now).

chu11 commented:
re-pushed, re-doing this PR based on RFC27 discussion in https://github.com/flux-framework/rfc/pull/251

chu11 commented:
rebased and re-pushed, this is the full RFC27 fixup of annotations

garlick commented:
Great, I'll make a review pass!

chu11 commented on src/modules/job-manager/alloc.c:
```diff
@@ -232,6 +234,43 @@ int cancel_request (struct alloc *alloc, struct job *job)
+                    (void)json_object_del (job->annotations, key);
+            }
+        }
+        /* Special case: if user cleared all entries, assume we no
```
i just realized, this should go inside the branch statement above this.  Got a fixup for it already.

garlick commented on src/common/libschedutil/alloc.c:
```diff
@@ -33,24 +33,18 @@ int schedutil_alloc_request_decode (const flux_msg_t *msg,
 static int schedutil_alloc_respond (flux_t *h, const flux_msg_t *msg,
                                     int type, const char *note,
-                                    const char *annotations_json_str)
+                                    json_t *annotations)
```
Since this refactors the code that was modified earlier, maybe ithe two libschedutil commits should be merged together?

chu11 replied:
i think I see what you mean, some of the changes could be done in an earlier commit to avoid "churn".  Don't think they can be fully merged together though.

garlick replied:
Looking at the code I attached this to it may not be clear - I mainly meant the changes to the  `schedutil_alloc_respond()` internal function.

garlick commented on src/modules/sched-simple/sched.c:
```diff
@@ -177,6 +177,41 @@ static char *Rstring_create (struct rlist *l, double now, double timelimit)
+    char *annotations_str = NULL;
+    char *rv = NULL;
+
+    if (!(annotations = json_object ())) {
+        flux_log_error (h, "json_object");
+        goto error;
+    }
+
+    if (!(jstr = json_string (str))) {
+        flux_log_error (h, "json_string");
+        goto error;
+    }
+
+    if (json_object_set (annotations, "sched.resource_summary", jstr) < 0) {
+        flux_log_error (h, "json_object_set");
+        goto error;
+    }
+
```
replace with json_pack?

chu11 replied:
oh duh :-)

chu11 commented on src/modules/job-manager/alloc.c:
```diff
@@ -238,33 +238,35 @@ static void update_annotations (flux_t *h, struct job *job, flux_jobid_t id,
+        if (!json_object_size (job->annotations)) {
+            json_decref (job->annotations);
+            job->annotations = NULL;
         }
```
yup, I caught that right after I pushed :P

chu11 commented on src/modules/job-manager/alloc.c:
```diff
@@ -34,7 +34,7 @@
  * Where type is one of:
  * 0 - resources allocated (sched commits R to KVS before responding)
- * 1 - annotation (just updates note, see below)
+ * 1 - annotation (updates note, see below)
```
ahh that's a good idea, many changes were originally made pre-RFC27.

garlick replied:
Was this forgotten?

chu11 replied:
do you have to refresh github?  It should be gone.

garlick replied:
Er.  Maybe.  My bad.

chu11 commented on t/job-manager/sched-annotate.c:
```diff
@@ -0,0 +1,321 @@
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* Simple scheduler for testing annotations:
+ * - presume that each job is requesting one core
+ * - track core counts, not specific core id's
+ * - mode=unlimited
+ *
+ * Command line usage:
+ *   flux module load sched-annotate [--cores=N]
+ * Options
+ *   --cores=N      specifies the total number of cores available (default 16)
+ */
+
```
let me see if that's more easily doable now, it was tougher before.  Basically the big change is supporting "single" and "unlimited".

grondo commented:
Running `t2204-job-manager-annotatesched.t` with `FLUX_TEST_VALGRIND` blows up. Not a critical issue since it seems to be in the test scheduler, but I thought I'd point it out.

```
expecting success: 
	flux job cancel $(cat job1.id) &&
	flux job cancel $(cat job3.id) &&
	flux job cancel $(cat job4.id) &&
	flux job cancel $(cat job5.id)

==7170== Thread 12:
==7170== Invalid read of size 8
==7170==    at 0x129A2E9C: exception_cb (sched-annotate.c:146)
==7170==    by 0x129A3C91: cancel_cb (ops.c:111)
==7170==    by 0x4E59B76: call_handler (msg_handler.c:231)
==7170==    by 0x4E5A240: dispatch_message (msg_handler.c:267)
==7170==    by 0x4E5A240: handle_cb (msg_handler.c:367)
==7170==    by 0x4E85052: ev_invoke_pending (ev.c:3372)
==7170==    by 0x4E88FF8: ev_run (ev.c:3775)
==7170==    by 0x4E58CF2: flux_reactor_run (reactor.c:126)
==7170==    by 0x129A35E4: mod_main (sched-annotate.c:318)
==7170==    by 0x114EBD: module_thread (module.c:214)
==7170==    by 0x59D26DA: start_thread (pthread_create.c:463)
==7170==    by 0x616088E: clone (clone.S:95)
==7170==  Address 0x84aa460 is 0 bytes inside a block of size 48 free'd
==7170==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==7170==    by 0x129A2DEA: job_destroy.part.0 (sched-annotate.c:58)
==7170==    by 0x129A2E77: job_destroy (stdio2.h:64)
==7170==    by 0x129A2E77: exception_cb (sched-annotate.c:150)
==7170==    by 0x129A3C91: cancel_cb (ops.c:111)
==7170==    by 0x4E59B76: call_handler (msg_handler.c:231)
==7170==    by 0x4E5A240: dispatch_message (msg_handler.c:267)
==7170==    by 0x4E5A240: handle_cb (msg_handler.c:367)
==7170==    by 0x4E85052: ev_invoke_pending (ev.c:3372)
==7170==    by 0x4E88FF8: ev_run (ev.c:3775)
==7170==    by 0x4E58CF2: flux_reactor_run (reactor.c:126)
==7170==    by 0x129A35E4: mod_main (sched-annotate.c:318)
==7170==    by 0x114EBD: module_thread (module.c:214)
==7170==    by 0x59D26DA: start_thread (pthread_create.c:463)
==7170==  Block was alloc'd at
==7170==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==7170==    by 0x129A31D2: job_create (sched-annotate.c:69)
==7170==    by 0x129A31D2: alloc_cb (sched-annotate.c:179)
==7170==    by 0x129A41B4: alloc_continuation (ops.c:41)
==7170==    by 0x4E85052: ev_invoke_pending (ev.c:3372)
==7170==    by 0x4E88FF8: ev_run (ev.c:3775)
==7170==    by 0x4E58CF2: flux_reactor_run (reactor.c:126)
==7170==    by 0x129A35E4: mod_main (sched-annotate.c:318)
==7170==    by 0x114EBD: module_thread (module.c:214)
==7170==    by 0x59D26DA: start_thread (pthread_create.c:463)
==7170==    by 0x616088E: clone (clone.S:95)
==7170== 
==7170== Invalid read of size 8
==7170==    at 0x4E5B197: flux_msg_get_type (message.c:376)
==7170==    by 0x4E5DD9A: request_decode (request.c:29)
==7170==    by 0x4E5E039: flux_request_vunpack (request.c:101)
==7170==    by 0x4E5E039: flux_request_unpack (request.c:119)
==7170==    by 0x129A43F5: schedutil_alloc_respond (alloc.c:41)
==7170==    by 0x129A2EA6: exception_cb (sched-annotate.c:146)
==7170==    by 0x129A3C91: cancel_cb (ops.c:111)
==7170==    by 0x4E59B76: call_handler (msg_handler.c:231)
==7170==    by 0x4E5A240: dispatch_message (msg_handler.c:267)
==7170==    by 0x4E5A240: handle_cb (msg_handler.c:367)
==7170==    by 0x4E85052: ev_invoke_pending (ev.c:3372)
==7170==    by 0x4E88FF8: ev_run (ev.c:3775)
==7170==    by 0x4E58CF2: flux_reactor_run (reactor.c:126)
==7170==    by 0x129A35E4: mod_main (sched-annotate.c:318)
==7170==  Address 0x8498c00 is 0 bytes inside a block of size 40 free'd
==7170==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==7170==    by 0x4E5AC61: flux_msg_destroy (message.c:226)
==7170==    by 0x129A2DE2: job_destroy.part.0 (sched-annotate.c:57)
==7170==    by 0x129A2E77: job_destroy (stdio2.h:64)
==7170==    by 0x129A2E77: exception_cb (sched-annotate.c:150)
==7170==    by 0x129A3C91: cancel_cb (ops.c:111)
==7170==    by 0x4E59B76: call_handler (msg_handler.c:231)
==7170==    by 0x4E5A240: dispatch_message (msg_handler.c:267)
==7170==    by 0x4E5A240: handle_cb (msg_handler.c:367)
==7170==    by 0x4E85052: ev_invoke_pending (ev.c:3372)
==7170==    by 0x4E88FF8: ev_run (ev.c:3775)
==7170==    by 0x4E58CF2: flux_reactor_run (reactor.c:126)
==7170==    by 0x129A35E4: mod_main (sched-annotate.c:318)
==7170==    by 0x114EBD: module_thread (module.c:214)
==7170==  Block was alloc'd at
==7170==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==7170==    by 0x4E5D5F1: flux_msg_create_common (message.c:188)
==7170==    by 0x4E5D5F1: flux_msg_copy (message.c:1465)
==7170==    by 0x129A3220: job_create (sched-annotate.c:76)
==7170==    by 0x129A3220: alloc_cb (sched-annotate.c:179)
==7170==    by 0x129A41B4: alloc_continuation (ops.c:41)
==7170==    by 0x4E85052: ev_invoke_pending (ev.c:3372)
==7170==    by 0x4E88FF8: ev_run (ev.c:3775)
==7170==    by 0x4E58CF2: flux_reactor_run (reactor.c:126)
==7170==    by 0x129A35E4: mod_main (sched-annotate.c:318)
==7170==    by 0x114EBD: module_thread (module.c:214)
==7170==    by 0x59D26DA: start_thread (pthread_create.c:463)
==7170==    by 0x616088E: clone (clone.S:95)

```

grondo commented on src/modules/job-manager/alloc.c:
```diff
@@ -276,14 +317,16 @@ static void alloc_response_cb (flux_t *h, flux_msg_handler_t *mh,
         if (event_job_post_pack (ctx->event, job, "alloc",
-                                 "{ s:s }",
-                                 "note", note ? note : "") < 0)
+                                 "{ s:O? }",
```
This will output a JSON `null` when `job->annotations` is `NULL`, is that what you intended? (Or was it meant to post an empty object? (Sorry if I missed something obvious)

chu11 replied:
yeah, that was intentional.  Would not outputting an annotation be preferable?  I was sort of going off the same pattern before, where an empty `note` was output with an empty string.

garlick replied:
[RFC 21](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_21.html#alloc-event) says annotations is an optional key, so it should probably be omitted if empty.

garlick commented:
Don't forget to update the title of this PR before it goes in.  Maybe "job-manager: support job annotations by scheduler"?

chu11 commented:
rebased and re-pushed.  This is a force push fixing up all of the things mentioned above.  Things are squashed since some of the changes were effectively  refactor how I did things before.  Key changes:

- instead of a new test scheduler `sched-annotate`, I added "unlimited" mode to `sched-dummy` to serve the same purpose
- fixed mem-leaks I didn't fix in `sched-annotate`.
- fixed up all the smaller things mentioned above

chu11 commented:
@dongahn we are getting closer to a PR that's worthy, so perhaps take a skim.  The big changes for `flux-sched` will be:

These functions have been renamed:

```
    schedutil_alloc_respond_note() to schedutil_alloc_respond_annotate()
    schedutil_alloc_respond_denied() to schedutil_alloc_respond_deny()
    schedutil_alloc_respond_R() to schedutil_alloc_respond_success()
```

and 

`schedutil_alloc_respond_annotate()`
`schedutil_alloc_respond_success()`

take json objects now (via encoded strings)

garlick commented:
> take json objects now (via encoded strings)

Not to make extra work, but thinking about @dongahn needing to make those changes, it's finally sinking in that you suggested a json_pack style interface here, and that was a really good idea, e.g.
```c
int schedutil_alloc_respond_annotate_pack (schedutil_t *util, const flux_msg_t *msg,
                                           const char *fmt, ...);

int schedutil_alloc_respond_success (schedutil_t *util, const flux_msg_t *msg,
                                     const char *R);
int schedutil_alloc_respond_success_pack (schedutil_t *util, const flux_msg_t *msg,
                                          const char *R);
                                          const char *fmt, ...);
```
Is it too late for this PR?

chu11 commented:
> Not to make extra work, but thinking about @dongahn needing to make those changes, it's finally sinking in that you suggested a json_pack style interface here, and that was a really good idea, e.g.

It shouldn't be hard to add.  I can just add it as a commit on top of everything else.

garlick commented:
> It shouldn't be hard to add. I can just add it as a commit on top of everything else.

Er, if it changes the interfaces (as I was assuming above) then it probably needs to be squashed with earlier commits that change those interfaces, and there would be a ripple effect in other commits...

chu11 commented:
> Er, if it changes the interfaces (as I was assuming above) then it probably needs to be squashed with earlier commits that change those interfaces, and there would be a ripple effect in other commits...

Oh, I see in your example above that you recommend removing the encoded json-strings option.  I was going to add an additional `_pack` function as an alternate option for users.

In your case, it'd be best to squash it in earlier.  It shouldn't be too bad.  I'll take a crack and see how bad  it is :P

chu11 commented:
```
int schedutil_alloc_respond_success (schedutil_t *util, const flux_msg_t *msg,
                                     const char *R);
int schedutil_alloc_respond_success_pack (schedutil_t *util, const flux_msg_t *msg,
                                          const char *R);
                                          const char *fmt, ...);
```

What's the point of having `schedutil_alloc_respond_success()` in your example above.  Wouldn't "NULL" for the `fmt` be sufficient?

chu11 commented:
> Er, if it changes the interfaces (as I was assuming above) then it probably needs to be squashed with earlier commits that change those interfaces, and there would be a ripple effect in other commits...

Thinking about this, playing with it a bit, I think it should be a commit on top now.  The current commits are more about supporting the annotations, including changing job-manager, RPC protocol, etc..

The `pack` functions are more about changing the API in `libschedutil` to be more convenient, which I think it s separate thing.

garlick commented:
> What's the point of having schedutil_alloc_respond_success() in your example above. Wouldn't "NULL" for the fmt be sufficient?

I guess that's fine.  (I can't say why it seems odd to set the fmt=NULL)

> The pack functions are more about changing the API in libschedutil to be more convenient, which I think it a separate thing.

If you think that makes sense i'll go with it!

chu11 commented:
Converting to pack functions was actually really easy to change. The change is:

```
    schedutil_alloc_respond_annotate() to schedutil_alloc_respond_annotate_pack()
    schedutil_alloc_respond_success() to schedutil_alloc_respond_success_pack()
```

threw a commit on top of the PR.

garlick commented on t/job-manager/sched-helper.sh:
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+#
+
+# job-manager sched helper functions
+
+get_state() {
+	local id=$1
+	local state=$(${LIST_JOBS} | awk '$1 == "'${id}'" { print $2; }')
+	test -z "$state" \
```
Suggestion given that these are separated from the test script:
* define LIST_JOBS in `sched-helper.sh` rather than depending on it being defined externally.
* consider more descriptive names and flagging functions that are not for external use like `get_state`
* add comment for each function explaining what it does and how to call it
* document that the annotation functions require HAVE_JQ in callers

grondo replied:
I was actually going to ask the `for try` loops in here are really necessary?
It would be good to understand why they are needed (perhaps it is obvious, sorry), since it indicates a potential use case for synchronizing with jobs in which there is a known race condition.

In this case `get_state` is synchronizing with the `clean` event and fetching job state directly from the job-manager (AFAICS), so the need for the try loop isn't clear.

In `check_annotation` it is perhaps a little more obvious why a try loop is required, since an annotation update can come at any time, so there is no "event" on which to synchronize. However, for future editors of the testsuite, a comment may not hurt. ;-)

chu11 replied:
> define LIST_JOBS in sched-helper.sh rather than depending on it being defined externally.

Thanks missed that!

chu11 replied:
> In this case get_state is synchronizing with the clean event and fetching job state directly from the job-manager (AFAICS), so the need for the try loop isn't clear.

TBH I didn't know either, as it was the code that was there beforehand.  I suspect the loop is there just in case the state transition happens a little slowly (although there should be a sleep for 0.5 in there, I'll add it).

Going through logs, it appears the `clean` event check is in  95bddec68de67145fe20c9bda70d9013fb1ba894.  If the job goes inactive and is no longer listed by the job-manager, the logic is to report the job state as inactive.  I'll add a comment to that effect.

garlick replied:
Yeah, I was thinking we might want to revisit these tests, but I wasn't sure I understood all the details (even though they are my creation from the early days of the job manager - sorry).  For now some comments would be a big improvement over what I left you with.

grondo replied:
Ok, I didn't realize these were not new functions. Sorry about that!

garlick replied:
There _are_ new functions, but they mimic the old ones.  I just didn't want to ask @chu11 to clean up my mess, though he's welcome to :-)

garlick commented on t/job-manager/sched-helper.sh:
```diff
@@ -0,0 +1,46 @@
+}
+
+check_annotation() {
+	local id=$1
+        local key=$2
+        local value="$3"
+	for try in $(seq 1 10); do
+		test "$(get_annotation $id $key)" = "${value}" && return 0
+                sleep 0.5
+	done
```
Seems like we have some inconsistent tab/space usage going on (judging from the diff)

chu11 commented on src/common/libschedutil/alloc.c:
```diff
@@ -156,15 +163,18 @@ static void alloc_continuation (flux_future_t *f, void *arg)
-    if (!(ctx = alloc_create (msg, R, annotations_json_str)))
+    va_start (ap, fmt);
+    if (!(ctx = alloc_create (msg, R, fmt, ap)))
         return -1;
```
thanks!  and I did not notice that `return -1` instead of `goto error`, so there was a potential missed `va_end()` call too.

chu11 commented:
repushed given comments above.  It was hard to do fixups without causing massive conflicts, so I interactively rebased and fixed things along the way.  So all fixes are already squashed.

For `sched-helper`, I renamed `no_annotation` to `check_no_annotations`, so that's a bit more consistent naming wise.  Put an underscore on the internal functions, to differentiate them a bit.  Hope that's ok.  Then added comments.  Note the extra commit https://github.com/flux-framework/flux-core/pull/2960/commits/eade2c54203259df607026336f508472ab6a79e1 to add the missing sleep in the try loop.

Also fixed up the `exception_cb` -> `cancel_cb` that was missed earlier.  Hope I did it right.  it's the last commit.

garlick commented:
Thanks!

> Also fixed up the exception_cb -> cancel_cb that was missed earlier. Hope I did it right. it's the last commit.

Just had a quick peek there - was hoping we could drop the "unused" args from the exported API signature since we're making a breaking change for schedulers in this PR (may as well do it now)

chu11 commented:
> Just had a quick peek there - was hoping we could drop the "unused" args from the exported API signature since we're making a breaking change for schedulers in this PR (may as well do it now)

Ahh I had missed that.  Correct the error in `sched-dummy` AND remove the unused args.

chu11 commented:
Threw another commit on top to remove the unused args out of the cancel callback.  The previous cleanup of `exception_cb` to `cancel_cb` in `sched-dummy` has been moved earlier in the commit series to be along with all of the other "cleanup" commits.

chu11 commented:
Awesome.  @dongahn ball is in your court now.  Hopefully you've seen my "make it compile" PR https://github.com/flux-framework/flux-sched/pull/681 to give you guys a head start on `flux-sched` fixes against this change.

dongahn commented:
> Awesome. @dongahn ball is in your court now. Hopefully you've seen my "make it compile" PR flux-framework/flux-sched#681 to give you guys a head start on flux-sched fixes against this change.

Thanks. Just came back from a short vacation and I'm catching up.

chu11 commented:
re-pushed with two extra commits, which will get squashed / re-arranged later.  One is to add the clearing of annotations, per discussion in #2608 and per recent RFC27 update (https://github.com/flux-framework/rfc/pull/254).  Then second commit is to add a test for it.

chu11 commented:
temporarily setting WIP - randomly trying some other pylint versions to see if the pylint problem goes away

chu11 commented:
removed WIP & re-pushed - the last 2 commits are the new additions.  I will squash and re-order the commits later before we merge.

chu11 commented:
> Last two commits look ok to me! Let's get this merged already! (Set MWP when you've squashed)

I think we need to wait for @dongahn to signal his ok, b/c of the `flux-sched` side.  Although now that I know https://github.com/flux-framework/flux-sched/issues/687 has been hit by other people, there's an increasing liklihood that my patch in https://github.com/flux-framework/flux-sched/pull/681 can be accepted (with a few minor tweaks).

garlick commented:
Oh right, sorry, forgot about that.  Had a panic when I thought this was still waiting on me.

chu11 commented:
> Oh right, sorry, forgot about that. Had a panic when I thought this was still waiting on me.

@dongahn gave his blessing for this to go in in https://github.com/flux-framework/flux-sched/pull/681

Since the `flux-sched` check will not pass above b/c of lib breakage, I assume someone has to manually hit the merge button?

chu11 commented:
@Mergifyio rebase

grondo commented:
Ok, if this one is ready I'll have to remove branch protections, then merge, then re-establish branch protections.

Good to merge?

chu11 commented:
Thanks @grondo
